lib/acme/client.rb

```diff
@@ -41,93 +39,221 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
+    payload = { certificate: base64_der_certificate }
+    payload[:reason] = reason unless reason.nil?
+
+    response = post(endpoint_for('revokeCert'), payload: payload)
+    response.success?
+  end
+
+  def get_nonce
+    response = Faraday.head(endpoint_for('newNonce'), nil, 'User-Agent' => USER_AGENT)
+    nonces << response.headers['replay-nonce']
+    true
   end
 
   private
 
+  def post(url, payload: {}, mode: :kid)
+    connection = connection_for(url: url, mode: mode)
+    connection.post(url, payload)
+  end
+
+  def get(url, mode: :kid)
+    connection = connection_for(url: url, mode: mode)
+    connection.get(url)
+  end
+
+  def download(url)
+    connection = connection_for(url: url, mode: :download)
+    connection.get do |request|
+      request.url(url)
+      request.headers['Accept'] = CONTENT_TYPES.fetch(:pkix)
+    end
+  end
+
+  def connection_for(url:, mode:)
+    uri = URI(url)
+    endpoint = "#{uri.scheme}://#{uri.hostname}:#{uri.port}"
+    @connections ||= {}
+    @connections[mode] ||= {}
+    @connections[mode][endpoint] ||= new_connection(endpoint: endpoint, mode: mode)
+  end
+
+  def new_connection(endpoint:, mode:)
+    Faraday.new(endpoint, **@connection_options) do |configuration|
+      configuration.use Acme::Client::FaradayMiddleware, client: self, mode: mode
+      configuration.adapter Faraday.default_adapter
+    end
+  end
+
   def fetch_chain(response, limit = 10)
     links = response.headers['link']
     if limit.zero? || links.nil? || links['up'].nil?
       []
     else
-      issuer = connection.get(links['up'])
+      issuer = get(links['up'])
       [OpenSSL::X509::Certificate.new(issuer.body), *fetch_chain(issuer, limit - 1)]
     end
   end
 
-  def load_directory!
-    @operation_endpoints = if @directory_uri
-      response = connection.get(@directory_uri)
-      body = response.body
-      {
-        'new-reg' => body.fetch('new-reg'),
-        'new-authz' => body.fetch('new-authz'),
-        'new-cert' => body.fetch('new-cert'),
-        'revoke-cert' => body.fetch('revoke-cert')
-      }
-    else
-      DIRECTORY_DEFAULT
+  def endpoint_for(key)
+    load_directory! unless @directory_loaded
+    @directory.fetch(key) do |missing_key|
+      raise Acme::Client::Error::UnsupporedOperation,
+        "Directory at #{@directory_url} does not include `#{missing_key}`"
     end
   end
+
+  def load_directory!
+    return false if @directory_loaded
+    body = fetch_directory
+    @meta = body['meta']
+
+    @directory = {}
+    @directory['newNonce'] = URI(body['newNonce']) if body['newNonce']
+    @directory['newAccount'] = URI(body['newAccount']) if body['newAccount']
+    @directory['newOrder'] = URI(body['newOrder']) if body['newOrder']
+    @directory['newAuthz'] = URI(body['newAuthz']) if body['newAuthz']
+    @directory['revokeCert'] = URI(body['revokeCert']) if body['revokeCert']
+    @directory['keyChange'] = URI(body['keyChange']) if body['keyChange']
```

Comment:
Maybe a loop here?

Reply:
You could also move this to a `Directory` class that has methods that return URIs for each of these operations. It could also encapsulate the logic in `endpoint_for` and `fetch_directory`

Reply:
Move it in an object. I was not sure but its pretty nice actually.